### DISCUSSION TITLE
Running models

### DISCUSSION THREAD

**Discussion by imenelydiaker** [embeddings-benchmark/mteb#705] [General]
Should we start running models for the paper?

We can define the scope here:

1. Number of models to run and their type (small, large ones, should also consider closed-source models).
3. Any idea of how to aggregate more than 443 tasks results in a nice readable table/figure? (open question on how to show the results)
4. Are we keeping [mteb/results](https://huggingface.co/datasets/mteb/results)? The only objection I see for merging it within the current repo is that _it could lead to errors_: contributors may delete or update files unintentionally and we may miss it during reviews as it gets lost between code files.
5. Where should we put the scripts that we'll use to run experiments?
6. Following the discussion here https://github.com/embeddings-benchmark/mteb/discussions/703 : it would be really nice if we could estimate the carbon footprint of our experiments. 

Any other considerations before starting the experiments?
@Muennighoff @KennethEnevoldsen @isaac-chung @x-tabdeveloping @orionw

**Comment by x-tabdeveloping** [comment]
Some models we should probably have:
 - `all-MiniLM-*-v2`
 - `paraphrase-multilingual-*`
 - `all-mpnet-*`
 - `multilingual-e5-*`
 - `e5-mistral-7b-instruct` (as it was the first to use Mistral+instructions for embedding)
 - `voyage-large-2-instruct` (as it is first on the MTEB leaderboard right now)
 - `text-embedding-large-3` (due to its relevance in the industry)
 - `GritLM-7b` & `LLM2Vec` supervised and unsupervised (due to methodological novelty and usefulness for creating new models)

Anything else comes to mind?

**Reply by imenelydiaker** [reply]
`all-MiniLM-*-v2` are English models, shouldn't it be better to focus on multilingual ones? (the `paraphrase-multilingual-*` ones).

Should we add `bge` models from `BAAI`? Although I think they're specialized in retrieval and reranking, so not relevant for other tasks.

**Reply by orionw** [reply]
+1 to the original list and @imenelydiaker's comment about a strong multilingual 7B model, if there is one...  

Would be nice to see some strong rerankers (mT5 based or SentenceTransformer rerankers) for retrieval tasks, but that would be in a sub-table.

**Reply by imenelydiaker** [reply]
@orionw sorry, I removed my comment, we may have not used it correctly for French, even though Mistral-7b is trained on mostly English texts. We can still keep `e5-mistral-7b-instruct` and I guess we also have `GritLM-7b`, should be enough for 7b models.

**Reply by imenelydiaker** [reply]
> Would be nice to see some strong rerankers (mT5 based or SentenceTransformer rerankers) for retrieval tasks, but that would be in a sub-table.

@orionw you mean adding a reranking step after the retrieval one before the evaluation? Else, we can have specialized models on retrieval and reranking tasks only, no need to evaluate them on other tasks.

**Reply by orionw** [reply]
Yes, sorry, I should use the term cross-encoders instead to disambiguate.  We have them added to MTEB now and I know they're very common for retrieval tasks.

**Reply by imenelydiaker** [reply]
Some popular cross-encoders:

- [bge-reranker-large](https://huggingface.co/BAAI/bge-reranker-large) and base (supports English and Chinese)
- [bge-reranker-v2-gemma](https://huggingface.co/BAAI/bge-reranker-v2-gemma) (Multilingual, based on gemma-2b)
- Cohere Reranker
- JinaAI rerankers: https://huggingface.co/jinaai/jina-reranker-v1-turbo-en, https://huggingface.co/jinaai/jina-reranker-v1-tiny-en (English only, see if it's worth to evaluate on other languages)

**Reply by KennethEnevoldsen** [reply]
I would add LASER to the list as well.

**Reply by wissam-sib** [reply]
Here are some other models I can think of:
* https://huggingface.co/sentence-transformers/LaBSE
* sentence-transformers/distiluse-base-multilingual-*
* xlm-roberta-base-multilingual-*
* https://huggingface.co/nomic-ai/nomic-embed-text-v1
* BAAI/bge-m3

Would be nice to also run one of the best english models (and a few others monolingual models for the most common languages).

**Reply by bp-high** [reply]
How many maximum models are gonna be run btw? Also should models trained on a family on languages (Indic languages, European languages) be tested to see how well they perform on the whole benchmark and the languages they were trained on?

A few language family specific models that come to my mind:
* l3cube-pune/indic-sentence-similarity-sbert (Indic Languages)
* symanto/sn-xlm-roberta-base-snli-mnli-anli-xnli (European + Asian languages)
* clips/mfaq (European languages)

**Reply by Ab1992ao** [reply]
Hi there, we are the team who create russian MTEB part. Our opinion mono lang models are important for evaluation, because they metrics often have higher scores that multiling models. Our team  takes a models aboard to eval on ru part of mteb:

ai-forever/sbert_large_nlu_ru
ai-forever/sbert_large_mt_nlu_ru
cointegrated/rubert-tiny
cointegrated/rubert-tiny2

We are reserve it for ours team, thanks.

**Reply by KennethEnevoldsen** [reply]
Sounds wonderful @Ab1992ao

**Comment by KranthiGV** [comment]
Can volunteers help with running the models?

**Reply by y12uc231** [reply]
I can help with that. Is there doc or any process written about which models to run?

**Reply by KennethEnevoldsen** [reply]
We have a few issues for tasks that needs to be sped up beforehand: #738 , #732, #660. It would be lovely if you have the time to help with these.

@Muennighoff for running the models do we need model implementation as well?We have a few issues for tasks that needs to be sped up beforehand: #738 , #732, #660. It would be lovely if you have the time to help with these.

@Muennighoff for running the models I assume we need model implementation as well?

**Reply by Muennighoff** [reply]
Yes I think we need the model implementations. I'd probably still put them in a new repo due to various pip requirements some models may have (e.g. `pip install gritlm` for GritLM etc)

**Reply by x-tabdeveloping** [reply]
We could just have optional imports in the implementations I think, and add them as optional dependencies, or just add helpful messages for the user to install the important packages.

**Reply by KennethEnevoldsen** [reply]
I will have the results standardized in an upcoming PR. Once this is in I believe we can start running the classification tasks.


@Muennighoff if people want to help here what is the best approach, to create the model implementations?

**Reply by Muennighoff** [reply]
Yes; I think we can also let other people run things if they want to and submit their results via PR?
(before we start running though we need a clear table of what models x what datasets to run though I think, maybe in a Google Sheet or markdown file and then people can add their name when they start running sth to avoid duplicate running)

**Reply by KennethEnevoldsen** [reply]
Yep. We would need to compare speed though but we can just do that on a sample task.

I believe we should run all tasks for `mteb.get_tasks()` (by default ignores superseeded tasks). At least for a representative set of models (to e.g. get correlation to figure out which to remove).

**Reply by imenelydiaker** [reply]
For the total of carbon emission reporting we'll probably need to run everything on the same hardware.. Else we can just mention the carbon footprint of tasks that we did speed, it should concern only some classification and clustering tasks.

**Reply by Muennighoff** [reply]
> For the total of carbon emission reporting we'll probably need to run everything on the same hardware

Why? Can't we just make sure everyone run with tracking turned on and then sum up the carbon emissions from every result file?

**Reply by mrshu** [reply]
> we'll probably need to run everything on the same hardware

No matter what the conclusion of this debate would be, I am happy to run at least a substantial set of tasks on a couple of H100s, if that could be of help.

**Reply by imenelydiaker** [reply]
> Why? Can't we just make sure everyone run with tracking turned on and then sum up the carbon emissions from every result file?

Yeah I guess we can do that if it's just reporting how much it did cost for the paper.

I guess I just had a different perspective, I thought we would like to give an estimation to users about how much running this benchmark would cost: e.g., running this benchmark on an A100 emitted XXX CO2eq.
And I realized it is too complicated because it depends on the energy type (renewable, coal, etc. according to the country) so nevermind, we'll just sum up everything and stick with reporting how much it did cost.

@mrshu you can run the experiment for the English benchmark: https://github.com/embeddings-benchmark/mteb/discussions/703#discussioncomment-9474411, we need your results to show that lighter benchmark -> better for the planet.

**Reply by mrshu** [reply]
Thanks @imenelydiaker, I am running the English benchmark (of the old, slow type) as we speak and will update the thread with the results once I get them.

**Reply by y12uc231** [reply]
hi @imenelydiaker , could you please share any steps or scripts you are using to run models? I am quite new to this repo so a reference or any sequence of steps would help me run some models as well. Thanks!

**Reply by AlexeyVatolin** [reply]
Hello, @imenelydiaker! Do you have a list of datasets for which we need to calculate the score?

**Reply by imenelydiaker** [reply]
Hi @y12uc231 and @AlexeyVatolin, 
We have an issue opened here about what's missing before we can start running the models: https://github.com/embeddings-benchmark/mteb/issues/784. We'll open a discussion soon with the set of experiments we're expecting and a guide.
Thank you for your patience!

**Reply by Ruqyai** [reply]
I would like to help and if there are steps that we follow, that would be great

**Reply by mrshu** [reply]
This will sound crazy, but the benchmark has been running for a week on a single H100 with `sentence-t5-xxl` and still has not finished. Having inspected the results closely, it seems we are in the middle of the Retrieval tasks right now, where each one takes about an hour. If all goes well, we should be done tomorrow.

**Reply by KennethEnevoldsen** [reply]
thanks @mrshu this indeed seems to be problematic. When you share the scores we should be able to inspect where the most promising speed-ups are.

**Reply by mrshu** [reply]
Just so that I don't forget, here are the results for the `sentence-t5-xxl` model: some of them are CO2 numbers are indeed crazy large: https://github.com/embeddings-benchmark/mteb/compare/main...mrshu:mteb:mrshu/port-carbon-emissions-estimation

**Reply by mrshu** [reply]
Just a small ping here -- happy to run some more experiments if it would help!
